### DISCUSSION TITLE
building a mp version to freeze some extra modules

### DISCUSSION THREAD

**Discussion by beetlegigg** [micropython/micropython#15502] [General]
Following my notes, garnered from where I'm not sure, my process starts with:
git clone https://github.com/micropython/micropython.git --branch master

This seems to grab the latest mp code as the build I did recently reports the mp on my picoW is:
MicroPython v1.24.0-preview.106.g55e75c4ad.dirty on 2024-07-15; Raspberry Pi Pico W with RP2040

But what I would really like to do is to include my extra modules into my version of 
[v1.23.0 (2024-06-02) .uf2](https://micropython.org/resources/firmware/RPI_PICO_W-20240602-v1.23.0.uf2)

Is this possible, and if so, can anyone point me to where/how to git clone the stable release for my nefarious manipulations. (or perhaps I'm not understanding the mp version that gets reported when one does a self build ?)

Thanks for any enlightenment that may be bestowed :-)

**Comment by shariltumin** [comment]
There are several ways to accomplish this. These are the steps I use to include freeze modules in my firmware build:

1. ```$ cd micropython```
2. ```$ mkdir extra```
3. ```$ cd extra```
4. ```$ vi mymod.py```
5. ```$ cd ../ports/rp2/boards```
6. ```$ cp -r RPI_PICO MY_RP2```
7. ```$ cd MY_RP2```
8. ```$ vi mpconfigboard.cmake```
9. ```$ vi manifest.py```

The "mpconfigboard.cmake":
```
# cmake file for Raspberry Pi Pico
set(PICO_BOARD "pico")

set(MICROPY_FROZEN_MANIFEST ${MICROPY_BOARD_DIR}/manifest.py)
```
Point the frozen manifest to the file in your board definition directory.

The "manifest.py"
```
freeze("$(PORT_DIR)/modules")
# include("$(MPY_DIR)/extmod/asyncio")
require("onewire")
require("ds18x20")
# require("dht")
require("neopixel")

# my extra module
freeze('$(MPY_DIR)/extra', 'mymod.py')
# freeze('$(MPY_DIR)/extra', 'mymod_1.py')
```

Include or exclude modules by commenting them out.

Then you can build your firmware in the ports/rp2 :
```bash
make V=1 BOARD=MY_RP2 submodules
make V=1 BOARD=MY_RP2 clean
make V=1 BOARD=MY_RP2
```

**Comment by beetlegigg** [comment]
@shariltumin 
Thanks for your info and your steps are well explained.  But do you have any insight as to what source file I'm actually getting when I download with the following command: 
**git clone https://github.com/micropython/micropython.git --branch master**

Am I getting the very latest nightly source code or 'preview' as as its now called - i.e. the source for the likes of:
[v1.24.0-preview.119.g8159dcc27 (2024-07-20) .uf2]

or am I getting the source code for the latest stable release - i.e. the source for:
[v1.23.0 (2024-06-02) .uf2]

The build I created a couple of days ago, in which I included some extra modules in the manifest, shows the following:
v1.24.0-preview.106.g55e75c4ad.dirty on 2024-07-15
which seems to indicate my build is based on the nightly build at that time.   

I was looking to check if my build was indeed based on a preview, or if it is based on a stable release, but, as it is a self build, thats just the way the build process names the uf2 file it produces. 

If it is based on a preview release, then my query seeks to know if there is a way to get the source for the stable release upon which to base my self build mp.  

But again thanks for your notes of building mp as your process notes are better than those I had stashed away. 👍

**Reply by shariltumin** [reply]
For version v1.23.0 you can try

```
$ git clone https://github.com/micropython/micropython --branch v1.23.0
$ cd micropython
$ git describe
v1.23.0

``` 
but the master branch (latest nightly) is probably the best.

**Reply by robert-hh** [reply]
If you clone the whole repository, you get **all** versions since the first commit. Then you can switch between the versions with e.g.

git checkout v1.23.0

For that one or 

git checkout master

For the latest one. For updates, you just have to run

git pull

It is advisable to update the submodule when checking out a different version to keep track of submodule changes, like with

git submodule sync;git submodule update --init

**Comment by beetlegigg** [comment]
Thanks to @shariltumin  and @robert-hh  for the knowledge on how to pick a version of mp source to download, that was the info I needed.
